src/rviz/validate_quaternions.h

```diff
@@ -40,104 +40,146 @@
+                        "Magnitude: %.3f", x, y, z, w, std::sqrt(norm2) );
+  return is_normalized;
 }
 
-inline bool normalizeQuaternion(double& w, double& x, double& y, double& z)
+inline double normalizeQuaternion( double& w, double& x, double& y, double& z )
 {
-  float norm2 = quaternionNorm2(w, x, y, z);
-  if (norm2 < 10e-3f)
-    return false;
-  norm2 = std::sqrt(norm2);
+  if ( 0.0 == x && 0.0 == y && 0.0 == z && 0.0 == w )
+  {
+    w = 1.0;
+    x = y = z = 0.0;
+    return 0.0;
+  }
+  double norm2 = quaternionNorm2( w, x, y, z );
+  norm2 = std::sqrt( norm2 );
   w /= norm2;
   x /= norm2;
   y /= norm2;
   z /= norm2;
-  return true;
+  return norm2;
 }
 
 
-inline bool validateQuaternion(const Ogre::Quaternion& quaternion)
+inline bool validateQuaternions( const Ogre::Quaternion& quaternion )
 {
-  return validateQuaternion(quaternion.w, quaternion.x, quaternion.y, quaternion.z);
+  return validateQuaternions( quaternion.w, quaternion.x, quaternion.y, quaternion.z );
 }
 
-inline bool validateQuaternion(const tf::Quaternion& quaternion)
+inline bool validateQuaternions( const tf::Quaternion& quaternion )
 {
-  return validateQuaternion(quaternion.w(), quaternion.x(), quaternion.y(), quaternion.z());
+  return validateQuaternions( quaternion.w(), quaternion.x(), quaternion.y(), quaternion.z());
 }
 
-inline bool validateQuaternion(const geometry_msgs::Quaternion &msg)
+inline bool validateQuaternions( const geometry_msgs::Quaternion &msg )
 {
-  return validateQuaternion(msg.w, msg.x, msg.y, msg.z);
+  return validateQuaternions( msg.w, msg.x, msg.y, msg.z );
 }
 
-inline bool validateQuaternion(const geometry_msgs::Pose &msg)
+inline bool validateQuaternions( const geometry_msgs::Pose &msg )
 {
-  return validateQuaternion(msg.orientation);
+  return validateQuaternions( msg.orientation );
 }
 
-inline bool validateQuaternion(const geometry_msgs::PoseStamped &msg)
+inline bool validateQuaternions( const geometry_msgs::PoseStamped &msg )
 {
-  return validateQuaternion(msg.pose.orientation);
+  return validateQuaternions( msg.pose );
 }
 
-
-inline bool normalizeQuaternion(Ogre::Quaternion& quaternion)
+inline double normalizeQuaternion( Ogre::Quaternion& quaternion )
 {
-  if (normalizeQuaternion(quaternion.w, quaternion.x, quaternion.y, quaternion.z))
-    return true;
-
-  quaternion.w = 1.0;
-  quaternion.x = quaternion.y = quaternion.z = 0.0;
-  return false;
+  return normalizeQuaternion( quaternion.w, quaternion.x, quaternion.y, quaternion.z );
 }
 
-inline bool normalizeQuaternion(const geometry_msgs::Quaternion &msg, Ogre::Quaternion &q)
+inline double normalizeQuaternion( const geometry_msgs::Quaternion &msg, Ogre::Quaternion &q )
 {
   q.w = msg.w;
   q.x = msg.x;
   q.y = msg.y;
   q.z = msg.z;
-  return normalizeQuaternion(q);
+  return normalizeQuaternion( q );
 }
 
-
 template<typename T>
-inline bool validateQuaternions(const T &vec)
+inline bool validateQuaternions( const std::vector<T> &vec )
```

Comment:
I more like the old version as it is more compact and easier to grasp ;-)

Reply:
yeah, I know, but I'm trying to keep unnecessary changes out for this PR 😃